docs/tries.md

```diff
@@ -0,0 +1,189 @@
+
+
+
+## Trie Data Structure: Explained with Code
+
+Here's a more detailed breakdown of the Trie implementation, with code snippets and explanations:
+
+**1. TrieNode Class:**
+
+```cpp
+#include <iostream>
+#include <map>
+#include <string>
+
+class TrieNode {
+public:
+  char data; // Character stored at this node
+  std::map<char, TrieNode*> children; // Map of child nodes
+
+  TrieNode(char ch) : data(ch) {} 
+};
+```
+* **`data`:**  Holds the character represented by this node.
+* **`children`:** A `std::map` is used to store child nodes, with the character leading to that child as the key. This efficiently manages the branching structure of the Trie. 
+**2. Trie Class:**
+```cpp
+class Trie {
+public:
+  TrieNode* root; // Root node of the Trie
+  Trie() {
+    root = new TrieNode('\0'); // Root node with null character
+  }
+  void insert(const std::string& word) {
+    TrieNode* current = root;
+    for (char ch : word) {
+      if (current->children.find(ch) == current->children.end()) {
+        current->children[ch] = new TrieNode(ch); // Create child node if not found
+      }
+      current = current->children[ch]; // Move to the child node
+    }
+  }
+  // ... other methods (search, startsWith, deleteString) 
+};
+```
+
+* **`root`:** Points to the root node of the Trie.
+* **`insert(std::string word)`:**
+    * Starts at the `root` node.
+    * Iterates through each character (`ch`) in the word:
+        * If a child node for `ch` doesn't exist, create one.
+        * Move to the corresponding child node.
+    * After processing all characters, mark the final node as the end of a word (optional, can be done with a flag or additional data in the node).
+
+**Example Usage:**
+
+```cpp
+int main() {
+  Trie trie;
+
+  trie.insert("apple");
+  trie.insert("banana");
+  trie.insert("app");
+
+  // Search for a word
+  if (trie.search("apple")) {
+    std::cout << "'apple' found in the Trie." << std::endl;
+  } 
+
+  return 0;
+}
+```
+
+**Explanation:**
+
+1. **Initialization:** The `Trie` object is created, initializing its `root` node with a null character.
+2. **Insertion:** The `insert` method is called for each word.  The Trie structure is built as described above.
+3. **Search:** The `search` method (not shown in the code snippet) would traverse the Trie based on the word's characters. If a path leading to the end of the word is found, it indicates the word exists in the Trie.
+
+**Key Points:**
+
+* **Time Complexity:**  Prefix searches in a Trie have a time complexity of O(k), where k is the length of the prefix.
+* **Memory Efficiency:**  Tries excel at storing words with common prefixes, significantly reducing storage space compared to storing individual strings.
+
+**Example (Visual):**
+
+Let's say we insert "apple" and "app":
+
+```
+     (Root)
+      | 
+      a 
+      |
+     p
+     |
+     p 
+     |
+     l
+     |
+     e 
+     | 
+     (word end)
+    /
+   p 
+   |
+   (word end)
+```
```

Comment:
Try Mermaid Graph for better understanding and good UI/UX. It follows best practices, and you can check out https://mermaid.js.org/ for reference.